### ISSUE TITLE
Correct setup for showing linked device state

### ISSUE THREAD

**Issue by mikeypdev** [kuestess/homebridge-platform-insteonlocal#146]
I'm having trouble getting HomeBridge to reflect the state of a linked device. In this example, I have a wall switch and a plug-in on/off module that are linked to each other. The wall switch is the controller, so when I manually use the switch, both the switch light and plug-in module lamp simultaneously turn on or off. A typical Insteon use case, I believe.

However, in HomeBridge, the plug-in module never reflects its current state. Am I missing something in the config? Here's what I have now:

 ```
                {
                    "name": "Bedroom Overhead",
                    "deviceID": "5108c5",
                    "deviceType": "switch",
                    "dimmable": "no"
                },
                {
                    "name": "Bedroom Lamp",
                    "deviceID": "280435",
                    "deviceType": "dimmer",
                    "dimmable": "no",
                    "controllers": [
                        "5108c5"
                    ]
                },
```

In the HomeBridge logs, I reliably see "Got on event" and "Got off event" for Bedroom Overhead when I manually use the switch.

HomeBridge: 1.3.4
node.js: v14.17.2
homebridge-platform-insteonlocal: v0.4.18 
Platform: Raspbian manual install

**Comment by kuestess** [comment]
@mikeypdev There are a lot of similar, yet different :-), features that have been implemented for the plugin to support specific use-cases.  The controllers feature is meant to update the status of (for example) linked controller devices when you actuate that device.  In your case, if you manually turn on 'Bedroom Lamp' then it should go out and check/update the status of 'Bedroom Overhead', however it won't check controllers if you actuate 'Bedroom Overhead' because the controllers aren't defined for that device.  Probably the easiest solution is to define 'Bedroom Overhead' as a scene and put 'Bedroom Lamp' as a group member.  Easy peasy.  Let me know if that doesn't make sense!

**Comment by mikeypdev** [comment]
Ah, thanks @kuestess, I see that I was thinking about the controllers setting backwards. So when I want a controller to update the status of a linked controlled device, I should use the groupMembers property. Which only works for scenes?

I updated my settings to make Bedroom Overhead a scene, and sure enough, the Bedroom Lamp status now gets updated correctly when I manually use the Bedroom Overhead switch. However, I no longer see the status of the Bedroom Overhead switch itself. I figure I need to somehow make this switch both a "switch" and a "scene", but I'm still a bit confused as to how to config it.

Current settings:

```
               {
                   "name": "Bedroom Overhead",
                   "deviceID": "5108c5",
                   "deviceType": "scene",
                   "dimmable": "no",
                   "groupMembers": "280435"
               },
               {
                   "name": "Bedroom Lamp",
                   "deviceID": "280435",
                   "deviceType": "dimmer",
                   "dimmable": "no",
                   "controllers": [
                       "5108c5"
                   ]
               },
```

**Comment by kuestess** [comment]
@mikeypdev I've done this in my setup by adding the device itself and a scene device - creates some clutter and an extra device, but it does work.  Scenes are really intended to be configured with keypads to indicate status, but I think there's a way to get this to work with only a single device.  For your config above for 'Bedroom Overhead', add `"groupID"` and set that to the hub group number.  Also to 'Bedroom Overhead', add `"keypadbtn": "A"`.  This should set up the single 'Bedroom Overhead' device to send a group command to the hub and look at your switch device itself to reflect the status.  This is a bit convoluted, and I can't believe the situation hasn't come up before, but it should work.

**Comment by mikeypdev** [comment]
Thanks @kuestess , it's almost there. Your recommendations work for updating the Bedroom Overhead status, but now the Bedroom Lamp status doesn't update. In fact, if I revert the changes, it still won't update. :-) The debug output shows no references to the groupMembers device. Is there a cache that I should refresh somewhere?

```
                {
                    "name": "Bedroom Overhead",
                    "deviceID": "5108c5",
                    "deviceType": "scene",
                    "groupID": "1",
                    "keypadbtn": "A",
                    "dimmable": "no",
                    "groupMembers": "280435"
                },
                {
                    "name": "Bedroom Lamp",
                    "deviceID": "280435",
                    "deviceType": "dimmer",
                    "dimmable": "no",
                    "controllers": [
                        "5108c5"
                    ]
                },

```

I'm not clear on which groupID to use. When I look at the device Links in the Insteon UI page (awesome tool, BTW), I see that the Bedroom Overhead has two groups, 0 and 1. 0 is linked to the Hub (a 2242) as a responder, and group 1 is linked to the Hub and the Bedroom Lamp as a controller. For the Bedroom Overhead device scene settings, it shows one scene with Controller Group 1, where as expected, Bedroom Overhead is the controller and Bedroom Lamp is the responder. It looks like all of my linked devices have similar setups.

I also have a few scenes set up on the Hub itself, where the hub is the controller. The Bedroom Overhead and Lamp are not part of any of those scenes. Should I make a new hub scene for the linked Bedroom devices?

(and my apologies, I can read the code, but I don't have enough node.js experience to try stepping through it to help debug)

thanks!

**Comment by mikeypdev** [comment]
@kuestess I've fixed the issue on my end, and I suspect it's a bug. I discovered that groupMember status would refresh if I used the device Name instead of deviceID in the groupMember field. This worked because the getGroupMemberStatus function doesn't validate the deviceID if the groupMember device is Name.

deviceID validation can fail on line 2406 for two reasons:
`			var isDefined = _.contains(deviceIDs, deviceID, 0)`

The first failure case is if the deviceIDs array is undefined. This was this case for me, and the fix is to change line 2387 from:
`	var deviceIDs = platform.deviceIDs`
to:
`	var deviceIDs = self.platform.deviceIDs`

The second failure case is due to case comparison. All of the devices listed in the deviceIDs array are in upper-case. If a groupDevice ID is listed in lower-case, such as 1234ab, _.contains will return false. The workaround for this is to simply ensure that deviceIDs in groupMembers are always upper-case in config.json, or to add toUpperCase() to line 2406
`			var isDefined = _.contains(deviceIDs, deviceID.toUpperCase(), 0)`

**Comment by mikeypdev** [comment]
@kuestess I submitted PR #147 to fix the above validation issue, as well as another PR #148  to add a feature that I wanted (momentary scene toggles). I also have a fix for the original issue #146 I reported, where the fix is adding groupMember support to on/off devices. I'll wait on the PR for that one until you accept/reject my other PRs, to avoid any possible merge conflicts.

**Comment by kuestess** [comment]
@mikeypdev Thanks for the PR - case has been a pain in the rear as some hubs do all upper case, some are lower case.  Momentary is a nice feature as well!
